blockstore/splitstore/splitstore.go

```diff
@@ -569,110 +766,189 @@ func (s *SplitStore) compact(curTs *types.TipSet) {
+	defer markSet.Close() //nolint:errcheck
+	defer s.debug.Flush()
 
-	// 1. mark reachable cold objects by looking at the objects reachable only from the cold epoch
-	log.Infow("marking reachable cold blocks", "boundaryEpoch", boundaryEpoch)
+	// 1. mark reachable objects by walking the chain from the current epoch; we keep state roots
+	//   and messages until the boundary epoch.
+	log.Info("marking reachable objects")
 	startMark := time.Now()
 
-	boundaryTs, err := s.chain.GetTipsetByHeight(context.Background(), boundaryEpoch, curTs, true)
-	if err != nil {
-		return xerrors.Errorf("error getting tipset at boundary epoch: %w", err)
-	}
-
 	var count int64
-	err = s.chain.WalkSnapshot(context.Background(), boundaryTs, 1, s.skipOldMsgs, s.skipMsgReceipts,
-		func(cid cid.Cid) error {
+	err = s.walkChain(curTs, boundaryEpoch, true,
+		func(c cid.Cid) error {
+			if isFilCommitment(c) {
+				return errStopWalk
+			}
+
 			count++
-			return coldSet.Mark(cid)
+			return markSet.Mark(c)
 		})
 
 	if err != nil {
-		return xerrors.Errorf("error marking cold blocks: %w", err)
+		return xerrors.Errorf("error marking: %w", err)
 	}
 
-	if count > s.markSetSize {
-		s.markSetSize = count + count>>2 // overestimate a bit
-	}
+	s.markSetSize = count + count>>2 // overestimate a bit
+
+	log.Infow("marking done", "took", time.Since(startMark), "marked", count)
+
+	// begin transactional protection with concurrent marking and fetch references created while marking
+	txnRefs := s.beginTxnConcurrentMarking(markSet)
+
+	// 1.1 Update markset for references created during marking
+	if len(txnRefs) > 0 {
+		log.Infow("updating mark set for live references", "refs", len(txnRefs))
+		startMark = time.Now()
+		walked := cid.NewSet()
+		count = 0
 
-	log.Infow("marking done", "took", time.Since(startMark))
+		for c := range txnRefs {
+			if isFilCommitment(c) {
+				continue
+			}
+
+			mark, err := markSet.Has(c)
+			if err != nil {
+				return xerrors.Errorf("error checking markset for %s: %w", c, err)
+			}
+
+			if mark {
+				continue
+			}
+
+			err = s.walkObjectIncomplete(c, walked,
+				func(c cid.Cid) error {
+					if isFilCommitment(c) {
+						return errStopWalk
+					}
+
+					mark, err := markSet.Has(c)
+					if err != nil {
+						return xerrors.Errorf("error checking markset for %s: %w", c, err)
+					}
+
+					if mark {
+						return errStopWalk
+					}
+
+					// we also short-circuit on old blocks, as these can come from a network request
+					// and cause us to fail because we have purged its consistituents (or missing from
+					// the beginning in case of snapshot sync, e.g. parent message receipts or old messages)
+					// if these blocks are on our chain, they would have been marked but they might be
+					// from a fork.
+					//
+					// Ideally, we would have API options to preclude us from trcking references to such
+					// objects, but we don't so we have to do this check
+					isOldBlock, err := s.isOldBlockHeader(c, lookbackEpoch)
```

Comment:
This is OK for now but it's definitely a hack. I'm not sure we want API options: they're going to be close to impossible to apply correctly at all relevant places (e.g. imagine a StateCompute JSON-RPC call for a historical tipset -- we would need to percolate the "no tracking" option down to the blockstore used by the VM...).

Instead, I'd suggest using the `ExposedBlockstore` DI wiring here (which we need to extend to the JSON-RPC layer).

To avoid dirtying the API with options, one idea is to add a `func (s *Splitstore) WithTrackingOmitted() blockstore.Blockstore)` method that returns the union of the hot and cold stores, without any tracking logic.

We can then set this as the `ExposedBlockstore`, and off we go.

Reply:
Agreed and I certainly don't want to have this code here permanently.
I fully agree that doing it through the exposed blockstore is the right approach, and the simplified view sounds like a good approach.

Plan is to do this in a subsequent pr.

Reply:
Going further, I think we can do this with a trait that defines `Expose() Blockstore`.
The exposed blockstore implementation wont have to know about splitstore then and just check if it implements the trait.

Reply:
Follow up issue: https://github.com/filecoin-project/lotus/issues/6705

Reply:
So I killed the `isBlockHeader` hack in https://github.com/filecoin-project/lotus/pull/6474/commits/4f89d260b05e7da546b82b7e0e2f42e50851c3a3
It was moribund code anyway, and we ought to resolve #6705 before shipping v1.11.1